File: app/src/main/java/com/example/authenticator/AccountAuthenticatorService.java
package com.example.authenticator;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

import java.util.Objects;

import com.example.api.Api;

import static android.accounts.AccountManager.ACTION_AUTHENTICATOR_INTENT;

public class AccountAuthenticatorService extends Service {
	private AccountAuthenticator authenticator = null;
	Api api;

	@Override
	public IBinder onBind(Intent intent) {
		if (intent != null && Objects.equals(intent.getAction(), ACTION_AUTHENTICATOR_INTENT)) {
			return getAuthenticator().getIBinder();
		}

		return null;
	}

	private AccountAuthenticator getAuthenticator() {
		if (authenticator == null) {
			authenticator = new AccountAuthenticator(this, api);
		}

		return authenticator;
	}

	@Override
	public void onCreate() {
		super.onCreate();
	}

	@Override
	public void onDestroy() {
		authenticator = null;
		super.onDestroy();
	}
}

File: app/src/main/kotlin/com/example/services/authenticator/AccountAuthenticatorService.kt
package com.example.services.authenticator

import android.accounts.AccountManager.ACTION_AUTHENTICATOR_INTENT
import android.app.Service
import android.content.Intent
import android.os.IBinder
import com.example.api.gateway.Api

class AccountAuthenticatorService : Service() {
  private var api: Api? = null
  private var authenticator: AccountAuthenticator? = null

  override fun onBind(intent: Intent?): IBinder? {
    return if (intent != null && intent.action == ACTION_AUTHENTICATOR_INTENT) {
      getAuthenticator().iBinder
    } else null

  }

  private fun getAuthenticator(): AccountAuthenticator {
    if (authenticator == null) {
      authenticator = AccountAuthenticator(this, this.api!!)
    }

    return authenticator as AccountAuthenticator
  }

  override fun onDestroy() {
    authenticator = null
    super.onDestroy()
  }
}
